src/ServiceControl.Transports.AmazonSQS/QueueLengthProvider.cs

```diff
@@ -0,0 +1,132 @@
+            var queue = QueueNameHelper.GetSqsQueueName(metadataReport.LocalAddress);
+
+            queues.AddOrUpdate(endpointInputQueue, _ => queue, (_, currentQueue) =>
+            {
+                if (currentQueue != queue)
+                {
+                    sizes.TryRemove(currentQueue, out var _);
+                }
+
+                return queue;
+            });
+
+            sizes.TryAdd(queue, 0);
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, TaggedLongValueOccurrence metricsReport)
+        {
+            //HINT: ASQ  server endpoints do not support endpoint level queue length monitoring
+        }
+
+        public Task Start()
+        {
+            stop = new CancellationTokenSource();
+
+            pooler = Task.Run(async () =>
+            {
+                using (var client = new AmazonSQSClient())
+                {
+                    var cache = new QueueUrlCache(client);
+
+                    while (!stop.Token.IsCancellationRequested)
+                    {
+                        try
+                        {
+                            await FetchQueueSizes(cache, client).ConfigureAwait(false);
+
+                            UpdateQueueLengthStore();
+
+                            await Task.Delay(QueryDelayInterval);
+                        }
+                        catch (Exception e)
+                        {
+                            Logger.Error("Error querying sql queue sizes.", e);
+                        }
+                    }
+                }
+            });
+
+            return TaskEx.Completed;
+        }
+
+        public Task Stop()
+        {
+            stop.Cancel();
+
+            return pooler;
+        }
+
+        void UpdateQueueLengthStore()
+        {
+            var nowTicks = DateTime.UtcNow.Ticks;
+
+            foreach (var tableNamePair in queues)
+            {
+                store.Store(
+                    new[]{ new RawMessage.Entry
+                    {
+                        DateTicks = nowTicks,
+                        Value = sizes.TryGetValue(tableNamePair.Value, out var size) ? size : 0
+                    }},
+                    tableNamePair.Key);
+            }
+        }
+
+        Task FetchQueueSizes(QueueUrlCache cache, IAmazonSQS client) => Task.WhenAll(sizes.Select(kvp => FetchLength(kvp.Key, client, cache)));
+
+        async Task FetchLength(string queue, IAmazonSQS client, QueueUrlCache cache)
+        {
+            try
+            {
+                var attReq = new GetQueueAttributesRequest
+                {
+                    QueueUrl = await cache.GetQueueUrl(queue).ConfigureAwait(false)
+                };
+                attReq.AttributeNames.Add("ApproximateNumberOfMessages");
+                var response = await client.GetQueueAttributesAsync(attReq).ConfigureAwait(false);
+                sizes[queue] = response.ApproximateNumberOfMessages;
```

Comment:
What are the caveats here of using the indexer vs the AddOrUpdate methods?

Reply:
Indexer uses a blunt fast way of adding or updating by 

```c#
TryAddInternal(key, value, true, true, out _);
```

`AddOrUpdate` uses merge function (and a factory function), a bit heavier.